### DISCUSSION TITLE
API for Allowed Heath Tags

### DISCUSSION THREAD

**Discussion by Ampflower** [samolego/GolfIV#61] [Ideas]
With the introduction of #60, there's a couple questions in terms of how other mods should interact with GolfIV when they have mobs that have special effects with their health (e.g. wolf's tail, iron golem's breakage, wither's shield).

### Possible ways to go about this
1. Custom properties in `fabric.mod.json` that we can query for via Fabric Loader.
   - This allows us to use FabricLoader's API directly to query metadata.
   - However, this API isn't hotpluggable into GSON as is, as it reimplements a GSON-like interface rather than directly implementing it. There would need to be a secondary parser on top to handle the metadata.
   - This does let the mod implementing this to have a simple API hook in without adding extra files.
2. A special `golfiv.json` file that mods can embed into the root of their jar.
   - We'll have to fetch via the jar directly. Fabric Loader does also provide an API for this, although we need to parse ourselves.
   - We could easily parse it however we wanted as we're not bounded to Fabric API's representation of the metadata.
   - This does require the mod implementing this to add extra files dedicated to GolfIV.

- Is there any other alternative that could be used?
- How would settings be layered for first time generation?

**Comment by samolego** [comment]
I think the `fabric.mod.json` option is much cleaner, paired with optional config support makes it really neat. I guess that config loading method could take it into account separately?

**Comment by samolego** [comment]
Just another thought, we could also have an api interface that would be implemented on all living entities, later overriden in iron golem, wither and wolf https://github.com/samolego/GolfIV/blob/61fe5f2e8684ddb2d768d8f2e79af3149c86c830/src/main/java/org/samo_lego/golfiv/storage/GolfConfig.java#L133

and having a default value as well.

If mod authors would like to change the behaviour, they'd just implement the interface on their entity and override the method?

**Reply by Ampflower** [reply]
That is a decent option, although there's the caveat of the interface always needing to be present, meaning there would need to be a detached API that can be JIJ'd.
